File: src/internal/client/policies.go
package client

import (
	"context"
	"fmt"
)

type PolicySettings struct {
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
	Visibility  string `json:"visibility,omitempty"`
}

type PolicyCreateRequest struct {
	UUID     string         `json:"uuid"`
	Settings PolicySettings `json:"settings"`
}

type PolicyUpdateRequest struct {
	UUID     string         `json:"uuid,omitempty"`
	Settings PolicySettings `json:"settings"`
}

type PolicyCreateResponse struct {
	PolicyID   int    `json:"policy_id"`
	PolicyName string `json:"policy_name"`
}

type PolicyDetail struct {
	ID                   int    `json:"id"`
	UUID                 string `json:"uuid"`
	Name                 string `json:"name"`
	Description          string `json:"description"`
	Owner                string `json:"owner"`
	OwnerID              int    `json:"owner_id"`
	Visibility           string `json:"visibility"`
	CreationDate         int    `json:"creation_date"`
	LastModificationDate int    `json:"last_modification_date"`
	NoTarget             string `json:"no_target"`
	TemplateUUID         string `json:"template_uuid"`
}

type PolicyDetailsResponse struct {
	UUID     string         `json:"uuid"`
	Settings PolicySettings `json:"settings"`
}

type PoliciesListResponse struct {
	Policies []PolicyDetail `json:"policies"`
}

func (c *Client) CreatePolicy(ctx context.Context, req PolicyCreateRequest) (*PolicyCreateResponse, error) {
	var resp PolicyCreateResponse
	if err := c.Post(ctx, "/policies", req, &resp); err != nil {
		return nil, fmt.Errorf("creating policy: %w", err)
	}
	return &resp, nil
}

func (c *Client) GetPolicy(ctx context.Context, policyID int) (*PolicyDetail, error) {
	var resp PolicyDetail
	if err := c.Get(ctx, fmt.Sprintf("/policies/%d", policyID), &resp); err != nil {
		return nil, fmt.Errorf("getting policy: %w", err)
	}
	return &resp, nil
}

func (c *Client) UpdatePolicy(ctx context.Context, policyID int, req PolicyUpdateRequest) error {
	if err := c.Put(ctx, fmt.Sprintf("/policies/%d", policyID), req, nil); err != nil {
		return fmt.Errorf("updating policy: %w", err)
	}
	return nil
}

func (c *Client) DeletePolicy(ctx context.Context, policyID int) error {
	if err := c.Delete(ctx, fmt.Sprintf("/policies/%d", policyID)); err != nil {
		return fmt.Errorf("deleting policy: %w", err)
	}
	return nil
}

func (c *Client) ListPolicies(ctx context.Context) (*PoliciesListResponse, error) {
	var resp PoliciesListResponse
	if err := c.Get(ctx, "/policies", &resp); err != nil {
		return nil, fmt.Errorf("listing policies: %w", err)
	}
	return &resp, nil
}

File: mockapi/tenableio_mock/routers/policies.py
"""Scan policies.

Three different shapes are in play and they do not match each other, which is
faithful to production:

* ``POST /policies`` takes ``{uuid, settings{...}}`` and returns only
  ``{policy_id, policy_name}``.
* ``GET /policies/{id}`` returns a flat detail object.
* ``GET /policies`` returns a list of those detail objects.
* ``PUT /policies/{id}`` returns an empty body.

The ``uuid`` on the request is the *template* the policy derives from, not the
policy's own identifier.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Request, Response

from ..errors import bad_request, not_found
from ..models import PolicyCreate, PolicyUpdate
from ..store import Store, unix_seconds
from ._common import get_settings, get_store, reject_unknown, resolve_on_omit

router = APIRouter(tags=["policies"])

MAX_DESCRIPTION = 3000
VISIBILITIES = {"private", "shared"}


def _validate(name: str, description: str | None, visibility: str | None) -> None:
    if not name or not name.strip():
        raise bad_request("The policy name is required.")
    if description is not None and len(description) > MAX_DESCRIPTION:
        raise bad_request(
            f"The policy description exceeds the maximum length of "
            f"{MAX_DESCRIPTION} characters."
        )
    if visibility is not None and visibility not in VISIBILITIES:
        raise bad_request(
            f"visibility must be one of: {', '.join(sorted(VISIBILITIES))}."
        )


def _find(store: Store, policy_id: int) -> dict[str, Any]:
    policy = store.policies.get(policy_id)
    if policy is None:
        raise not_found(f"Policy {policy_id} was not found.")
    return policy


@router.post("/policies")
async def create_policy(body: PolicyCreate, request: Request) -> dict[str, Any]:
    store, settings = get_store(request), get_settings(request)
    reject_unknown(body, settings)
    _validate(body.settings.name, body.settings.description, body.settings.visibility)
    if not body.uuid or not body.uuid.strip():
        raise bad_request("The template uuid is required.")

    with store.lock:
        stamp = unix_seconds(store.now())
        policy = {
            "id": store.next_id(),
            # The policy gets its own uuid; body.uuid identified the template.
            "uuid": store.next_uuid(),
            "name": body.settings.name,
            "description": body.settings.description or "",
            "owner": settings.user,
            "owner_id": 1,
            "visibility": body.settings.visibility or "private",
            "creation_date": stamp,
            "last_modification_date": stamp,
            "no_target": "false",
            "template_uuid": body.uuid,
        }
        store.policies[policy["id"]] = policy
        return {"policy_id": policy["id"], "policy_name": policy["name"]}


@router.get("/policies")
async def list_policies(request: Request) -> dict[str, Any]:
    store = get_store(request)
    with store.lock:
        return {"policies": [dict(p) for p in store.policies.values()]}


@router.get("/policies/{policy_id}")
async def get_policy(policy_id: int, request: Request) -> dict[str, Any]:
    store = get_store(request)
    with store.lock:
        return dict(_find(store, policy_id))


@router.put("/policies/{policy_id}")
async def update_policy(
    policy_id: int, body: PolicyUpdate, request: Request
) -> Response:
    store, settings = get_store(request), get_settings(request)
    reject_unknown(body, settings)
    _validate(body.settings.name, body.settings.description, body.settings.visibility)

    with store.lock:
        policy = _find(store, policy_id)
        policy["name"] = body.settings.name
        policy["description"] = resolve_on_omit(
            body.settings,
            "description",
            policy["description"],
            settings.quirks.on_omitted_description,
        )
        if body.settings.visibility is not None:
            policy["visibility"] = body.settings.visibility
        if body.uuid:
            policy["template_uuid"] = body.uuid
        policy["last_modification_date"] = unix_seconds(store.now())
    return Response(status_code=200)


@router.delete("/policies/{policy_id}")
async def delete_policy(policy_id: int, request: Request) -> Response:
    store = get_store(request)
    with store.lock:
        _find(store, policy_id)
        del store.policies[policy_id]
    return Response(status_code=200)
